docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -73,27 +74,57 @@ and includes all the core field members that are common across all transaction t
 Developers do not have to include `StdTxBase` if they wish, so it is meant to be
 used as an auxiliary type.
 
-### Signing
+### Signing & Transaction Submission
 
-Signing of a `Transaction` must be canonical across clients and binaries. In order
-to provide canonical representation of a `Transaction` to sign over, clients must
+To provide an acceptable user experience for client-side developers, we separate
+transaction encoding and signing. While every app defines its own transaction
+types for encoding purposes, we define a single standard signing `SignDoc` to be used
+across apps that leverages `google.protobuf.Any` to handle interface types. `Any`
+can be used to wrap any protobuf message by encoding both its type URL and value.
+We avoid using `Any` during the encoding phase because these URLs can be long
+and we want to keep encoded messages small. We are not, however, sending the
+`SignDoc` (encoded in JSON) over the wire or storing it on nodes,
+therefore using `Any` for signing does not add any encoding overhead.
+
+```proto
+message SignDoc {
+  StdSignDocBase base = 1;
+  repeated google.protobuf.Any msgs = 2;
+}
+```
+
+Signing of a `SignDoc` must be canonical across clients and binaries. In order
+to provide canonical representation of a `SignDoc` to sign over, clients must
 obey the following rules:
 
-- Encode `SignDoc` (see below) via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
+- Encode `SignDoc` via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
   - Default must be stripped from the output!
   - JSON keys adhere to their Proto-defined field names.
 - Generate canonical JSON to sign via the [JSON Canonical Form Spec](https://gibson042.github.io/canonicaljson-spec/).
   - This spec should be trivial to interpret and implement in any language.
 
-```Protobuf
-// app/codec/codec.proto
+Because signing and encoding are separated and apps will know how to encode
+a signing messages, we can provide a gRPC/REST endpoint that allows for generic
+app-independent transaction submission. This can be done via a generic
+`AnyTransaction` type:
 
-message SignDoc {
-  StdSignDocBase base = 1;
-  repeated Message msgs = 2;
+```proto
+message AnyTransaction {
```

Comment:
It's not clear to me what purpose `AnyTransaction` serves apart from being used during tx signing. How does this fit into tx generation and broadcasting?

Are you stating that module tx endpoints will generate tx JSON blobs using `AnyTransaction`. Will these JSON blobs then be fed to `/tx/broadcast`, where they're JSON decoded and then Proto re-encoded? 

Please make this very clear.

Reply:
This should be `SignDoc`+signatures, i.e. a signed transaction in an application agnostic container.

Reply:
Yes AnyTransaction can be used for tx broadcasting. It is related to SignDoc but not quite the same. This is basically the same exact design that we currently have with amino. It just involves a re-encoding step to translate Any to the oneof's